File: scenarios/haskell/hs-binary-parser/template/tests/TestBinaryParser.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import BinaryParser
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as BC
import Data.Bits
import Data.Word
import Data.Int
import System.Exit
import Control.Monad

-- Helper to pack big endian
packU32 :: Word32 -> [Word8]
packU32 n = map (fromIntegral . (.&. 0xFF) . shiftR n) [24, 16, 8, 0]

packU64 :: Word64 -> [Word8]
packU64 n = map (fromIntegral . (.&. 0xFF) . shiftR n) [56, 48, 40, 32, 24, 16, 8, 0]

main :: IO ()
main = do
    testLogin
    testOrder
    putStrLn "All tests passed."

testLogin :: IO ()
testLogin = do
    let magic = 0xAF :: Word8
    let ver = 1 :: Word8
    let typ = 1 :: Word8
    let userLen = 5 :: Word8
    let user = BC.unpack "admin"
    let userBytes = map (fromIntegral . fromEnum) user
    let hashBytes = replicate 32 0xAA :: [Word8]
    
    let payload = [magic, ver, typ, userLen] ++ userBytes ++ hashBytes
    let s = sum (map fromIntegral payload) :: Word32
    let chk = packU32 s
    
    let bs = B.pack (payload ++ chk)
    
    case parsePacket bs of
        Right (Login u h) -> do
            when (u /= "admin") $ die "Username mismatch"
            when (B.length h /= 32) $ die "Hash len mismatch"
            putStrLn "Test Login: PASSED"
        Right _ -> die "Wrong packet type"
        Left e -> die $ "Parse failed: " ++ e

testOrder :: IO ()
testOrder = do
    -- Simple mock, Double encoding is tricky manually without IEEE lib, 
    -- but let's assume 'parsePacket' does standard IEEE754.
    -- We'll skip complex double verification or just use a known bit pattern for 1.0 or similar.
    -- 150.50 is approx ... let's use 0.0 for simplicity or just check other fields primarily.
    -- Better: Assume the parser expects standard 8-bytes. We pass 8 bytes logic.
    let magic = 0xAF :: Word8
    let ver = 1 :: Word8
    let typ = 2 :: Word8
    
    -- ID = 12345
    let pid = packU64 12345
    let sym = map (fromIntegral . fromEnum) "AAPL" -- 4 bytes
    let side = 1 :: Word8
    -- Price 0.0 (all zeros usually workable, or we can try to pack 150.50 manually? No.)
    -- Let's just use 8 zero bytes for price and assert it is 0.0.
    let price = replicate 8 0
    let qty = packU32 100
    
    let payload = [magic, ver, typ] ++ pid ++ sym ++ [side] ++ price ++ qty
    let s = sum (map fromIntegral payload) :: Word32
    let chk = packU32 s
    
    let bs = B.pack (payload ++ chk)
    
    case parsePacket bs of
         Right (Order i s si p q) -> do
            when (i /= 12345) $ die "ID mismatch"
            when (s /= "AAPL") $ die "Symbol mismatch"
            when (q /= 100) $ die "Qty mismatch"
            putStrLn "Test Order: PASSED"
         Right _ -> die "Wrong packet type"
         Left e -> die $ "Parse failed: " ++ e

File: scenarios/ocaml/ml-binary-parser/template/tests/test_binary_parser.ml
open BinaryParser
open Printf

let pack_u32 buf offset v =
  for i = 0 to 3 do
    Bytes.set_uint8 buf (offset + i) (Int32.to_int (Int32.logand (Int32.shift_right v ((3 - i) * 8)) 255l))
  done

let pack_u64 buf offset v =
  for i = 0 to 7 do
    Bytes.set_uint8 buf (offset + i) (Int64.to_int (Int64.logand (Int64.shift_right v ((7 - i) * 8)) 255L))
  done

let test_login () =
  let magic = 0xAF in
  let ver = 1 in 
  let typ = 1 in
  let user_len = 5 in
  let username = "admin" in
  let hash = Bytes.make 32 '\xAA' in
  
  let header_len = 3 in
  let payload_len = 1 + 5 + 32 in
  let total = header_len + payload_len + 4 in
  let buf = Bytes.create total in
  
  Bytes.set_uint8 buf 0 magic;
  Bytes.set_uint8 buf 1 ver;
  Bytes.set_uint8 buf 2 typ;
  Bytes.set_uint8 buf 3 user_len;
  Bytes.blit_string username 0 buf 4 5;
  Bytes.blit hash 0 buf 9 32;
  
  let sum = ref 0 in
  for i = 0 to header_len + payload_len - 1 do
    sum := !sum + (Bytes.get_uint8 buf i)
  done;
  
  (* Simple mod 2^32 logic actually implies wrapping sum *)
  let chk = Int32.of_int !sum in (* Assume it fits or wraps *)
  pack_u32 buf (header_len + payload_len) chk;
  
  match parse_packet buf with
  | Ok (Login {username=u; hash=h}) ->
      if u <> "admin" then failwith "Username mismatch";
      if Bytes.length h <> 32 then failwith "Hash len mismatch";
      printf "Test Login: PASSED\n"
  | Ok _ -> failwith "Wrong packet type"
  | Error e -> failwith ("Parse failed: " ^ e)

let test_order () =
  let magic = 0xAF in
  let ver = 1 in 
  let typ = 2 in
  
  let header_len = 3 in
  let id = 123456789012345L in
  let symbol = "AAPL" in (* 4 bytes *)
  let side = 1 in
  let price = 150.50 in
  let qty = 100l in
  
  let payload_len = 8 + 4 + 1 + 8 + 4 in
  let total = header_len + payload_len + 4 in
  let buf = Bytes.create total in
  
  Bytes.set_uint8 buf 0 magic;
  Bytes.set_uint8 buf 1 ver;
  Bytes.set_uint8 buf 2 typ;
  
  pack_u64 buf 3 id;
  Bytes.blit_string symbol 0 buf 11 4;
  Bytes.set_uint8 buf 15 side;
  (* Price is double. Pack logic is tricky. Let's assume user implementation handles Int64_bits of float *)
  let price_bits = Int64.bits_of_float price in
  (* Endianness: Int64.bits_of_float is host order, but prompt implies BigEndian standard protocol? 
     Usually binary parsers clarify. We'll use BigEndian for Int64. *)
     (* If host is Little Endian, bits_of_float returns LE int64. We might need to swap for BE? 
        Or just pack it as is if we expect the parser to read bits_of_float directly.
        Let's pack it as BigEndian Int64 bits. *)
  pack_u64 buf 16 price_bits; (* This puts it in BE if pack_u64 does BE *)
    
  pack_u32 buf 24 qty;
  
  let sum = ref 0 in
  for i = 0 to header_len + payload_len - 1 do
    sum := !sum + (Bytes.get_uint8 buf i)
  done;
  pack_u32 buf (header_len + payload_len) (Int32.of_int !sum);
  
  match parse_packet buf with
  | Ok (Order {id=i; symbol=s; price=p}) ->
      if i <> id then failwith "ID mismatch";
      if s <> symbol then failwith "Symbol mismatch";
      if abs_float (p -. price) > 0.001 then failwith "Price mismatch";
      printf "Test Order: PASSED\n"
  | Ok _ -> failwith "Wrong packet type"
  | Error e -> failwith ("Parse failed: " ^ e)

let () =
  try
    test_login ();
    test_order ();
    printf "All tests passed.\n"
  with Failure f ->
    printf "FAILED: %s\n" f;
    exit 1
